clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java

```diff
@@ -0,0 +1,411 @@
+                offsetFetcherUtils.updateSubscriptionState(multiNodeResult.fetchedOffsets,
+                        isolationLevel);
+
+                if (listOffsetsRequestState.remainingToSearch.size() == 0) {
+                    ListOffsetResult listOffsetResult =
+                            new ListOffsetResult(listOffsetsRequestState.fetchedOffsets,
+                                    listOffsetsRequestState.remainingToSearch.keySet());
+                    listOffsetsRequestState.globalResult.complete(listOffsetResult);
+                } else {
+                    requestsToRetry.add(listOffsetsRequestState);
+                }
+            } else {
+                log.debug("ListOffsets request failed with error", error);
+                listOffsetsRequestState.globalResult.completeExceptionally(error);
+            }
+        });
+
+        for (Map.Entry<Node, Map<TopicPartition, ListOffsetsRequestData.ListOffsetsPartition>> entry : partitionResetTimestampsByNode.entrySet()) {
+            Node node = entry.getKey();
+
+            CompletableFuture<ListOffsetResult> partialResult = sendListOffsetRequestToNode(
+                    node,
+                    entry.getValue(),
+                    requireTimestamps,
+                    unsentRequests);
+
+            partialResult.whenComplete((result, error) -> {
+                if (error != null) {
+                    multiNodeRequest.resultFuture.completeExceptionally(error);
+                } else {
+                    multiNodeRequest.addPartialResult(result);
+                }
+            });
+        }
+        return unsentRequests;
+    }
+
+    /**
+     * Build ListOffsetRequest to send to a specific broker for the partitions and
+     * target timestamps. This also adds the request to the list of unsentRequests.
+     **/
+    private CompletableFuture<ListOffsetResult> sendListOffsetRequestToNode(
+            Node node,
+            Map<TopicPartition, ListOffsetsRequestData.ListOffsetsPartition> targetTimes,
+            boolean requireTimestamps,
+            List<NetworkClientDelegate.UnsentRequest> unsentRequests) {
+        ListOffsetsRequest.Builder builder = ListOffsetsRequest.Builder
+                .forConsumer(requireTimestamps, isolationLevel, false)
+                .setTargetTimes(ListOffsetsRequest.toListOffsetsTopics(targetTimes));
+
+        log.debug("Creating ListOffsetRequest {} for broker {} to reset positions", builder,
+                node);
+
+        NetworkClientDelegate.UnsentRequest unsentRequest = new NetworkClientDelegate.UnsentRequest(
+                builder,
+                Optional.ofNullable(node));
+        unsentRequests.add(unsentRequest);
+        CompletableFuture<ListOffsetResult> result = new CompletableFuture<>();
+        unsentRequest.future().whenComplete((response, error) -> {
+            if (error != null) {
+                log.error("Sending ListOffsetRequest {} to broker {} failed",
```

Comment:
I see the offsetFetcher uses debug level for this message: mind double check?

Reply:
`log.debug("Sending ListOffsetRequest {} to broker {}", builder, node);`

Reply:
These 2 are not for the same case, that's why I introduced it differently, but totally agree with your point. I updated the log levels to align with the OffsetFetch approach, so sticking to debug level for all these internal flows.